plugins/github-enricher/sponsorFinder.js

```diff
@@ -0,0 +1,395 @@
+//const fetch = require("node-fetch")  // TODO is this needed when test mocks it?
+const promiseRetry = require("promise-retry")
+const PersistableCache = require("./persistable-cache")
+const yaml = require("js-yaml")
+
+// We store the raw(ish) data in the cache, to avoid repeating the same request multiple times and upsetting the github rate limiter
+const DAY_IN_SECONDS = 60 * 60 * 24
+const DAY_IN_MILLISECONDS = 1000 * DAY_IN_SECONDS
+
+const cacheOptions = { stdTTL: 4 * DAY_IN_SECONDS }
+const CACHE_KEY = "github-api-for-contributions"
+const COMPANY_CACHE_KEY = CACHE_KEY + "-company"
+const REPO_CACHE_KEY = CACHE_KEY + "-repo"
+
+const repoContributorCache = new PersistableCache(cacheOptions)
+const companyCache = new PersistableCache(cacheOptions)
+
+let minimumContributorCount = 2
+let minimumContributionPercent = 20
+
+// Only consider users who have a significant number of contributions
+let minContributionCount = 5
+
+let optedInSponsors
+
+const setMinimumContributionPercent = n => {
+  minimumContributionPercent = n
+}
+
+const setMinimumContributorCount = n => {
+  minimumContributorCount = n
+}
+
+const setMinimumContributionCount = n => {
+  minContributionCount = n
+}
+
+const initSponsorCache = (cache) => {
+  companyCache.ingestDump(cache.get(COMPANY_CACHE_KEY))
+  repoContributorCache.ingestDump(cache.get(REPO_CACHE_KEY))
+
+}
+
+const clearCaches = () => {
+  repoContributorCache.flushAll()
+  companyCache.flushAll()
+  optedInSponsors = undefined
+}
+
+const getSponsorData = async () => {
+  if (!optedInSponsors) {
+    const url = "https://raw.githubusercontent.com/quarkusio/quarkus-extension-catalog/main/named-contributing-orgs-opt-in.yml"
```

Comment:
Instead of grabbing from GitHub, how about cloning the repository and using the local file? That will get rid of network issues

Reply:
I see what you mean, but I'm not sure it will help that much. We only grab the file once, and grabbing the single file should more network-friendly than pulling down the whole repo. Even for the file access, I don't think we pass an our access token, so I don't think the rate limiter 'sees' us do it.

The rate limiter is kicking in later on, when we hammer the GitHub API with several queries per repository.

Reply:
> You don't actually need to clone the whole repo, you can `git sparse-checkout` only a directory if you want (eg. https://askubuntu.com/questions/460885/how-to-clone-only-some-directories-from-a-git-repository).

Good point! Although it still seems a bit like solving a problem that doesn't really exist. Doing anything on the command line with the `git` cli would add a lot of complexity, and adding a dependency on a node git library would make the dependency tree fatter. And if I make an http call to execute the clone, then I'm basically doing the same as what I do now.

> Also since this file rarely changes, do you think that grabbing the file during the website build may be a reasonable choice instead?

All of this code is 100% website-build-time. It's run to generate the static site that's pushed to GitHub pages (so we only hit the GitHub API when we do a GitHub actions build, not when a user visits the page). A lot of it is also (now) cached between builds, although I don't cache the sponsor information between builds since it would be annoying if people had to wait several days for their updates to reflect on the page.

(As a rough map, if it's in `gatsby-*` or `plugins` it's build-time, and if it's in `src` it's client-side and run-time.

Reply:
Ah, that makes sense. For some obscure reason, I thought that was used as a Javascript library when deployed. Thanks for the explanation!

Reply:
It does get confusing, because all the code looks the same but it's got two different lifecycle phases. I've sometimes caught myself doing expensive calculations in the runtime path, just because ... it's all javascript, all in the same repo. 

I sometimes wonder if we should do even more rendering build-time to make the site as snappy as possible. I think gatsby supports it, but I haven't looked into it. And I also wonder the opposite, and wonder if as the complexity of the front-end grows, if we need to be less reliant on build-time calculations and go to something more dynamic. For now there's probably not much in it either way, since I think the (current) front-end performance issues wouldn't be helped by more static-ness.

Reply:
I think it's fine to render more stuff build-time, given that the website is frequently built, that would save us from some headaches I think

Reply:
Network issues and reaching the rate-limit basically

Reply:
Ah, that's all build time at the moment. We don't do any network calls at run time. Even images are inlined and sprited and otherwise optimised, I think. All the rate limiting headaches are build time (and, hopefully, fixed now that I've got caching in place, except for this PR which is using an unwarmed-cache). 

Things that are slightly dynamic at the moment are, for example, is that when you load the front page, we have the names of all the extensions in memory and then dynamically decide which to show. That means there's no network call when you apply a filter, but it means a bit more processing on the client side. On the first call there's no filter, by definition, so some [Gatsby SSR](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-ssr/) might make load faster. There's also some React code that we might be able to streamline out ... but it's all pretty subtle stuff, and I think there's lower hanging fruit.